contrib/zip/example/main.cc

```diff
@@ -0,0 +1,159 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "contrib/zip/sandboxed.h"
+#include "contrib/zip/utils/utils_zip.h"
+
+ABSL_FLAG(bool, list, false, "list files");
+ABSL_FLAG(std::string, unzip, "", "unzip");
+ABSL_FLAG(std::string, add_file, "", "add file");
+ABSL_FLAG(std::string, delete, "", "delete");
+
+static absl::Status ListFiles(LibZip& zip) {
+  SAPI_ASSIGN_OR_RETURN(int64_t num, zip.GetNumberEntries());
+  for (uint64_t i = 0; i < num; i++) {
+    SAPI_ASSIGN_OR_RETURN(std::string name, zip.GetName(i));
+    std::cout << name << "\n";
+  }
+
+  return absl::OkStatus();
+}
+
+static absl::Status Unzip(LibZip& zip, std::string filename) {
+  SAPI_ASSIGN_OR_RETURN(std::vector<uint8_t> buf, zip.ReadFile(filename));
+
+  std::cout << buf.data();
+
+  return absl::OkStatus();
+}
+
+static absl::Status AddFile(LibZip& zip, std::string filename) {
+  int fd = open(filename.c_str(), O_RDONLY);
+  if (fd < 0) {
+    return absl::UnavailableError(
+        absl::StrCat("Unable to open file ", filename));
+  }
+
+  // The fd will be consumed
+  SAPI_ASSIGN_OR_RETURN(uint64_t index, zip.AddFile(filename, fd));
+
+  return absl::OkStatus();
+}
+
+static absl::Status DeleteFile(LibZip& zip, std::string filename) {
+  int64_t index = -1;
+
+  SAPI_ASSIGN_OR_RETURN(int64_t num, zip.GetNumberEntries());
+  for (uint64_t i = 0; i < num; i++) {
+    SAPI_ASSIGN_OR_RETURN(std::string name, zip.GetName(i));
+    if (name == filename) {
+      index = i;
+      break;
+    }
+  }
+  if (index == -1) {
+    return absl::UnavailableError(
+        absl::StrCat("Unable to remove file ", filename));
+  }
+
+  return zip.DeleteFile(index);
+}
+
+int main(int argc, char* argv[]) {
+  std::string prog_name(argv[0]);
+  google::InitGoogleLogging(argv[0]);
+  std::vector<char*> args = absl::ParseCommandLine(argc, argv);
+
+  if (args.size() < 2 || args.size() > 3) {
+    std::cerr << "Usage:\n  " << prog_name << " ZIPFILE [OUTFILE]\n";
+    return EXIT_FAILURE;
+  }
+
+  std::string filename(args[1]);
+  ZipSapiSandbox sandbox;
+  if (!sandbox.Init().ok()) {
+    std::cerr << "Unable to start sandbox\n";
+    return EXIT_FAILURE;
+  }
+
+  LibZip zip(&sandbox, filename, 0);
+
+  if (!zip.IsOpen()) {
+    std::cerr << "Unable to open file " << filename << "\n";
+    return EXIT_FAILURE;
+  }
+
+  int outfd = -1;
+  if (args.size() == 3) {
+    outfd = open(args[2], O_WRONLY | O_CREAT);
+    if (outfd < 0) {
+      std::cerr << "Unable to open file " << args[2] << "\n";
+      return EXIT_FAILURE;
+    }
+  }
+
+  bool modification = false;
```

Comment:
Optional: I think `needs_saving` is a better name.

Reply:
Fixed.